packages/runner-shared/src/event-manager.js

```diff
@@ -500,6 +518,87 @@ export const eventManager = {
         studioRecorder.testFailed()
       }
     })
+
+    Cypress.on('test:before:run', (...args) => {
+      Cypress.primaryOriginCommunicator.toAllSpecBridges('test:before:run', ...args)
+    })
+
+    Cypress.on('test:before:run:async', (...args) => {
+      Cypress.primaryOriginCommunicator.toAllSpecBridges('test:before:run:async', ...args)
+    })
+
+    // Inform all spec bridges that the primary origin has begun to unload.
+    Cypress.on('window:before:unload', () => {
+      Cypress.primaryOriginCommunicator.toAllSpecBridges('before:unload')
+    })
+
+    Cypress.primaryOriginCommunicator.on('window:load', ({ url }, originPolicy) => {
+      // Sync stable if the expected origin has loaded.
+      // Only listen to window load events from the most recent secondary origin, This prevents nondeterminism in the case where we redirect to an already
+      // established spec bridge, but one that is not the current or next cy.origin command.
+      if (cy.state('latestActiveOriginPolicy') === originPolicy) {
+        // We remain in an anticipating state until either a load even happens or a timeout.
+        cy.state('autOrigin', cy.state('autOrigin', cors.getOriginPolicy(url)))
+        cy.isAnticipatingCrossOriginResponseFor(undefined)
+        cy.isStable(true, 'load')
+        // Prints out the newly loaded URL
+        Cypress.emit('internal:window:load', { type: 'cross:origin', url })
+        // Re-broadcast to any other specBridges.
+        Cypress.primaryOriginCommunicator.toAllSpecBridges('window:load', { url })
+      }
+    })
+
+    Cypress.primaryOriginCommunicator.on('before:unload', () => {
+      // We specifically don't call 'cy.isStable' here because we don't want to inject another load event.
+      // Unstable is unstable regardless of where it initiated from.
+      cy.state('isStable', false)
+      // Re-broadcast to any other specBridges.
+      Cypress.primaryOriginCommunicator.toAllSpecBridges('before:unload')
+    })
+
+    Cypress.primaryOriginCommunicator.on('expect:origin', (originPolicy) => {
+      localBus.emit('expect:origin', originPolicy)
+    })
+
+    Cypress.primaryOriginCommunicator.on('viewport:changed', (viewport, originPolicy) => {
+      const callback = () => {
+        Cypress.primaryOriginCommunicator.toSpecBridge(originPolicy, 'viewport:changed:end')
+      }
+
+      Cypress.primaryOriginCommunicator.emit('sync:viewport', viewport)
+      localBus.emit('viewport:changed', viewport, callback)
+    })
+
+    Cypress.primaryOriginCommunicator.on('before:screenshot', (config, originPolicy) => {
+      const callback = () => {
+        Cypress.primaryOriginCommunicator.toSpecBridge(originPolicy, 'before:screenshot:end')
+      }
+
+      handleBeforeScreenshot(config, callback)
+    })
+
+    Cypress.primaryOriginCommunicator.on('url:changed', ({ url }) => {
+      localBus.emit('url:changed', url)
+    })
+
+    Cypress.primaryOriginCommunicator.on('after:screenshot', handleAfterScreenshot)
```

Comment:
Is the primaryOriginCommunicator different than Cypress.action?

Reply:
Yes, it's what manages communication between the primary driver and any secondary driver(s). It's an abstraction over postMessage.